### DISCUSSION TITLE
ImageSharp.Drawing transformations are unintuitive

### DISCUSSION THREAD

**Discussion by ryanackley** [SixLabors/ImageSharp.Drawing#82] [Ideas]
If you're coming from any other graphics API, ImageSharp's transformations are unintuitive because they behave a little strangely in comparison. Look at this code

```c#
transform.AppendTranslation(new SixLabors.Primitives.PointF(0, 2.75f * dpi));
transform.AppendRotationDegrees(slantedAngle);
img.Mutate(ctx =>
{
     ctx.Transform(transform);
     ctx.DrawText(text, font, SixLabors.ImageSharp.Color.Black, new SixLabors.ImageSharp.PointF(0f, 0f));
});
```
Coming from OpenGL, GDI+, Javascript Canvas, etc., you would expect this code to apply the transformation to `DrawText `. What it actually does is apply the transformation to what's already been drawn. This makes it impossible to apply discrete transformations on a single draw operation. For example, drawing a single thing rotated at an angle. I'm very new to ImageSharp so maybe I'm missing something.

Pretty much every other drawing library out there applies the transform as a state on the drawing context that applies to subsequent operations. You can typically manage the state using push/pop type functions.

**Comment by JimBobSquarePants** [comment]
Please do not open non-issues in the issue tracker. There is a template (that you deleted) for a reason.

I will transform this to the correct repository (drawing) and move it to the correct discussions channel (ideas) so that any conversation can continue there.

**Comment by ryanackley** [comment]
Thanks for being a maintainer on this library. I understand you're a volunteer. You don't owe me anything. Keep in mind, I'm volunteering my time too. By taking the time to create this issue and articulate my feedback and provide the benefit of my experience, I am trying to help this project. I have no other ulterior motive. We've already figured out workarounds and have working code. There is no expectation that this will be fixed to address a particular need. 

1. Didn't delete the template. Github didn't present me with one. Probably a bug because I wasn't logged in. It redirected me to authenticate after I clicked the create issue button. After I authenticated, I was presented with a blank create issue form.
2. When you moved the issue, Github didn't link from the old issue. It said the issue was moved to a repo I don't have permission on or was deleted. So I created another one. Wasn't trying to spam you.

**Reply by JimBobSquarePants** [reply]
Oh wow! Please accept my sincerest apologies. I just managed to recreate the GitHub issue. That's not great at all!

Regarding your input around transforms. There is a fundamental difference in the way this library works compared to say.. GDI+

With System.Drawing and GDI your transform is bound contextually to an instance of the `Graphics` class so any transforms are applied there.

Our processing context should not be seen as an equivalent to that class although it does allow us to set the state of certain properties to be shared across operations. Each chained operation applied within the `Mutate` or `Clone` method is a separate process which is equivalent to `Graphics` instance that can consume those shared properties.

``` c#
transform.AppendTranslation(new SixLabors.Primitives.PointF(0, 2.75f * dpi));
transform.AppendRotationDegrees(slantedAngle);
img.Mutate(ctx =>
{
     ctx.Transform(transform); // Transform the image
     ctx.DrawText(text, font, SixLabors.ImageSharp.Color.Black, new SixLabors.ImageSharp.PointF(0f, 0f)); // Draw text upon the transformed image
});
```

Now that all said. I do agree that it's no easy to apply a simple transform to text. It would be nice to be able to pass a matrix to the methods though I believe that would require changes to the Fonts library.

@tocsoft what do you think?

**Reply by tocsoft** [reply]
should be simple enough and no it would only require changes inside `ImageSharp.Drawing` its the rendering layer that would apply the transform not the layout.

**Reply by JimBobSquarePants** [reply]
Perfect. I'll create a tracking issue and we can flesh out the specification there.

**Reply by JimBobSquarePants** [reply]
Tracking issue created #83

**Reply by ryanackley** [reply]
It's totally possible the paradigm shift is going right over my head or it's just not clicking so ignore me if that's the case.

The text transform you propose would have solved the problem we ran into. We ended up creating glyphs along a path to rotate the text to a desired angle. The use case was a watermark on a financial instrument. 

I still think there is a very strong case to be made for transformation states because it provides a fine-grained way to apply discrete transforms on operations. Also, the mental model required to composite complex drawings is better served. For example, I wrote a PowerPoint slide rasterizer for a commercial product (Atlassian's Confluence) in 2007 using  java Drawing. A slide can have many different regions. Each region can have it's own transform, etc. Without transformation states, it becomes an absolute mess to follow and debug the conversion between local and world space of each "object" being rasterized. I'm sure it's only a matter of time before someone tries to write an SVG rasterizer for imagesharp. Seems like it would have the same challenges.

Another data point: for this particular project, we were creating a financial instrument defined by an ANSI spec. It has well-defined regions with blocks of text. They have to be at a very specific coordinates on the image with specific text. The way our guy did it with ImageSharp was to create individual regions as separate images then composite them together on a master image. Seems a little verbose to me but that may be the way you guys want the model to work. He could have acccomplished the same thing on one image with transformation states and not strained his math muscles too hard (which I assume is what he was trying to avoid). 

Anyway, that's my 2 cents, thanks for listening.

**Reply by JimBobSquarePants** [reply]
> I'm sure it's only a matter of time before someone tries to write an SVG rasterizer for imagesharp. Seems like it would have the same challenges.

Someone already has made a start on an SVG rasterizer and seems to be having very little issue.
https://github.com/UkooLabs/SVGSharpie

I think what you are missing is that each method is completely separate and imperative. They're not connected in any way.

>They have to be at a very specific coordinates on the image with specific text. The way our guy did it with ImageSharp was to create individual regions as separate images then composite them together on a master image. 

I don't understand why they would do this. You can already set the location of text.
https://docs.sixlabors.com/api/ImageSharp.Drawing/SixLabors.ImageSharp.Drawing.Processing.DrawTextExtensions.html#SixLabors_ImageSharp_Drawing_Processing_DrawTextExtensions_DrawText_IImageProcessingContext_SixLabors_ImageSharp_Drawing_Processing_TextGraphicsOptions_System_String_Font_Color_PointF_

**Reply by antonfirsov** [reply]
@ryanackley in other words: 
`IImageProcessingContext` wasn't designed to be a Canvas API for ImageSharp. It's just a "thing" that represents the image to be operated on, for the time a chain of image processing operations being executed on `Image`. It's has been deliberately kept free of the heavy state management you can find in GDI's `Graphics` and others.

@JimBobSquarePants @tocsoft on the other hand, I'm wondering if there is a valid user need to introduce a Canvas API within ImageSharp.Drawing to serve as "sugar" in scenarios like this:

```C#
abstract class DrawableSceneObject
{
    public Matrix Transform { get; }
    public Pen Pen { get; }
    public Brush Brush { get; }

    public void Draw(ICanvas c)
    {
        c.CurrentTransform = Transform;
        c.CurrentPen = Pen;
        c.CurrentBrush = Brush;

        DrawCore(c);
    }

    protected abstract DrawCore(ICanvas c)
}

class Line : DrawableSceneObject
{
    public Vector2 Start, End;
    protected override DrawCore(ICanvas c) => c.DrawLine(Start, End);
}
```

Draft of possible API:

```C#
public interface ICanvas : IImageProcessingContext {
    public Matrix3x3 CurrentTransform { get; set; }
    public Pen CurrentPen { get; set;}
    public Brush CurrentBrush { get; set; }
}


public static class ImageExtensions {    
    // Same as Mutate, but uses the stateful ICanvas
    public static Draw(this Image image, Func<ICanvas, ICanvas> drawFunc) { }
}

public static class DrawRectangleExtensions {
    public static ICanvas DrawRectangle(this ICanvas c, Rectangle r) => (ICanvas)c.DrawRectangle(r.Transform(c.CurrentTransform), c.CurrentPen, c.CurrentBrush);
}
```

**Reply by tocsoft** [reply]
I was planning on proposing and trying to shift us over to have this sort of API... We already have state in the `IImageProcessingContext` we don't need to create a new concept for this we just need to first provide overloads that take in a transformation Matrix (pre-req. for all follow on work) then on all operation call site update to load a default transform from the grab bag on `IImageProcessingContext` (if its set) and pass it down into the processors to intelligently apply as required... 

Same can happen for `Brush`s and `Pen`s too if required, however we would probably have to break the `DrawText` apis too accommodate that because we currently have different logic based on whether a pen/brush is passed in we would have to change that to separate APIs calls one for outline one for draw as we could have a transient pen around that probably wont want to be used by the text drawing code.

**Reply by antonfirsov** [reply]
@tocsoft ah I see, so we basically evolved to the Canvas-like design. It was clearly the goal with the property bag, but I somehow missed the point.

**Reply by antonfirsov** [reply]
I wish we could rename `IImageProcessingContext` to `IImageCanvas` now.

**Reply by JimBobSquarePants** [reply]
> I wish we could rename IImageProcessingContext to IImageCanvas now.

Nah @antonfirsov context is fine. Canvas to me suggest something I apply multiple layers to.

@tocsoft Are you thinking adding the property to the bag for the Drawing APIs only? Adding the the main would be confusing for consumers of the Transform extensions.

**Reply by JimBobSquarePants** [reply]
@tocsoft @antonfirsov I cannot believe it's been 4 years since we looked at this! 

I actually want to rejig ImageSharp.Drawing to introduce a `ICanvas` like type for the next major version. 

If we designed it correctly (completely separate from `IImageProcessingContext`) we could allow for different implementations (GPU) and potentially make the library a candidate for UI generation. 

My thought process is that it would work like System.Drawing `Graphics` or Skia's `SKCanvas`. 

What do you think?
